linera-core/src/client/mod.rs

```diff
@@ -1395,11 +1395,63 @@ where
                 return Err(error.into());
             }
         };
-        for received_certificates in responses {
-            // Process received certificates.
-            self.receive_certificates_from_validator(received_certificates)
-                .await;
+        let response_count = responses.len();
+        let mut other_sender_chains = BTreeSet::new();
+        let mut certificates = BTreeMap::<ChainId, BTreeMap<BlockHeight, Certificate>>::new();
+        for response in responses {
+            other_sender_chains.extend(response.other_sender_chains);
+            // Update the tracker: Even if the validator sent us unsuitable certificates, there is
+            // no need to ask them again.
+            self.state_mut()
+                .update_received_certificate_tracker(response.name, response.tracker);
+            for certificate in response.certificates {
+                certificates
+                    .entry(certificate.value().chain_id())
+                    .or_default()
+                    .insert(certificate.value().height(), certificate);
+            }
         }
+        let certificate_count = certificates.values().map(BTreeMap::len).sum::<usize>();
+        tracing::info!(
+            "Received {certificate_count} certificates from {response_count} validators."
+        );
+
+        let chain_worker_limit = CHAIN_WORKER_LIMIT.saturating_sub(3).max(1);
```

Comment:
Maybe document why this formula is used.

Reply:
I added a comment explaining this.